File: packages/models/src/Line.ts
import Coordinate, { type ICoordinateLiteral } from "./Coordinate.js";
import type { ITrail } from "./Trail.js";
import Trail from "./Trail.js";

/**
 * Line to model a drawn lat/lng line
 */
export interface ILine {
  id?: number;

  start: ICoordinateLiteral;
  end: ICoordinateLiteral;

  // Associations
  trail?: ITrail;
}

interface Line extends ILine {}

class Line {
  start: Coordinate;
  end: Coordinate;
  trail?: Trail;

  constructor(data: ILine) {
    this.id = data.id;
    this.start = Coordinate.build(data.start);
    this.end = Coordinate.build(data.end);
    this.trail = data.trail ? new Trail(data.trail) : undefined;
  }
}

export default Line;

File: go-server/models/line.go
package models

type Line struct {
	ID    *int       `json:"id,omitzero"`
	Start Coordinate `json:"start"`
	End   Coordinate `json:"end"`
	Trail *Trail     `json:"trail,omitzero"`
}

func (l Line) IsZero() bool {
	return l.ID == nil && l.Start.IsZero() && l.End.IsZero() && l.Trail == nil
}
